Atsumaru: Some chapter pages fail to load/download, `NXDOMAIN` error

### Source information

Atsumaru 1.4.9

### Source language

English

### Steps to reproduce

1. Install the Atsumaru extension
2. Go to certain chapters on certain manga and try to read them

Here are some example chapters which fail for me:
- Gokurakugai: 5-6.5, 11-13, 16, 20-21, 27, 30-31
- The Executed Sage Who Was Reincarnated as a Lich Started an All-Out War: 4-6, 12, 14, 18, 20-21, 23-24, 26, 29, 37, 44, 49, 52, 58
- An Observation Log of My Wife Who Calls Herself a Villainess: 2-3, 6, 8-9, 11, 14, 16
- MAD: 8, 11, 15, 27, 29, 33, 35, 37, 40-41
- The Elusive Samurai: 3-4, 18, 26, 29-30, 34, 36, 41, 44, 56-57, 59, 66, 74, 79, etc

### Expected behavior

Chapter loads / downloads normally

### Actual behavior

All images fail to load, when I try to download a chapter that isn't working I get an error message in server logs saying:
```Suwayomi-Server  | [DefaultDispatcher-worker-23] DEBUG suwayomi.tachidesk.manga.impl.download.Downloader source(2327480808438768017) - downloadChapter(23197 - 236692 | state= Queued, tries= 1, progress= 0.0)) -- start
Suwayomi-Server  | [DefaultDispatcher-worker-23] DEBUG suwayomi.tachidesk.manga.impl.chapter.ChapterForDownload(mangaId= 23197, chapterId= 236692, chapterIndex= 15)::asDownloadReady -- isMarkedAsDownloaded= false, dbPageCount= 40, downloadPageCount= 0
Suwayomi-Server  | [DefaultDispatcher-worker-23] DEBUG suwayomi.tachidesk.manga.impl.chapter.ChapterForDownload(mangaId= 23197, chapterId= 236692, chapterIndex= 15)::asDownloadReady -- reset download status and fetch page list
Suwayomi-Server  | [DefaultDispatcher-worker-23] DEBUG eu.kanade.tachiyomi.network.NetworkHelper -- --> GET https://atsu.moe/api/read/chapter?mangaId=jx1GA&chapterId=2iudPagS h2
Suwayomi-Server  | [DefaultDispatcher-worker-23] DEBUG eu.kanade.tachiyomi.network.NetworkHelper -- <-- 200 https://atsu.moe/api/read/chapter?mangaId=jx1GA&chapterId=2iudPagS (324ms, unknown-length body)
Suwayomi-Server  | [DefaultDispatcher-worker-23] WARN suwayomi.tachidesk.manga.impl.download.Downloader source(2327480808438768017) - downloadChapter(23197 - 236692 | state= Queued, tries= 1, progress= 0.0)) -- failed due to
Suwayomi-Server  | java.io.IOException: atsu.moehttps
Suwayomi-Server  | 	at eu.kanade.tachiyomi.source.online.HttpSource.getImage$suspendImpl(HttpSource.kt:374)
Suwayomi-Server  | 	at eu.kanade.tachiyomi.source.online.HttpSource.getImage(HttpSource.kt)
Suwayomi-Server  | 	at suwayomi.tachidesk.manga.impl.Page$getPageImage$6.invokeSuspend(Page.kt:138)
Suwayomi-Server  | 	at suwayomi.tachidesk.manga.impl.Page$getPageImage$6.invoke(Page.kt)
Suwayomi-Server  | 	at suwayomi.tachidesk.manga.impl.Page$getPageImage$6.invoke(Page.kt)
Suwayomi-Server  | 	at suwayomi.tachidesk.manga.impl.util.storage.ImageResponse.getImageResponse(ImageResponse.kt:69)
Suwayomi-Server  | 	at suwayomi.tachidesk.manga.impl.Page.getPageImage(Page.kt:137)
Suwayomi-Server  | 	at suwayomi.tachidesk.manga.impl.Page.getPageImage$default(Page.kt:57)
Suwayomi-Server  | 	at suwayomi.tachidesk.manga.impl.Page.getPageImageDownload(Page.kt:188)
Suwayomi-Server  | 	at suwayomi.tachidesk.manga.impl.download.fileProvider.ChaptersFilesProvider.downloadImpl(ChaptersFilesProvider.kt:156)
Suwayomi-Server  | 	at suwayomi.tachidesk.manga.impl.download.fileProvider.ChaptersFilesProvider.access$downloadImpl(ChaptersFilesProvider.kt:73)
Suwayomi-Server  | 	at suwayomi.tachidesk.manga.impl.download.fileProvider.ChaptersFilesProvider$download$1.execute(ChaptersFilesProvider.kt:218)
Suwayomi-Server  | 	at suwayomi.tachidesk.manga.impl.download.fileProvider.ChaptersFilesProvider$download$1.execute(ChaptersFilesProvider.kt:218)
Suwayomi-Server  | 	at suwayomi.tachidesk.manga.impl.ChapterDownloadHelper.download(ChapterDownloadHelper.kt:46)
Suwayomi-Server  | 	at suwayomi.tachidesk.manga.impl.download.Downloader.run(Downloader.kt:147)
Suwayomi-Server  | 	at suwayomi.tachidesk.manga.impl.download.Downloader.access$run(Downloader.kt:38)
Suwayomi-Server  | 	at suwayomi.tachidesk.manga.impl.download.Downloader$run$1.invokeSuspend(Downloader.kt)
Suwayomi-Server  | 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:34)
Suwayomi-Server  | 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:98)
Suwayomi-Server  | 	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:586)
Suwayomi-Server  | 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:829)
Suwayomi-Server  | 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:717)
Suwayomi-Server  | 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:704)
Suwayomi-Server  | Caused by: java.net.UnknownHostException: atsu.moehttps
Suwayomi-Server  | 	at java.base/java.net.InetAddress$CachedLookup.get(Unknown Source)
Suwayomi-Server  | 	at java.base/java.net.InetAddress.getAllByName0(Unknown Source)
Suwayomi-Server  | 	at java.base/java.net.InetAddress.getAllByName(Unknown Source)
Suwayomi-Server  | 	at okhttp3.Dns$Companion$DnsSystem.lookup(Dns.kt:50)
Suwayomi-Server  | 	at okhttp3.internal.connection.RouteSelector.resetNextInetSocketAddress(RouteSelector.kt:170)
Suwayomi-Server  | 	at okhttp3.internal.connection.RouteSelector.nextProxy(RouteSelector.kt:132)
Suwayomi-Server  | 	at okhttp3.internal.connection.RouteSelector.next(RouteSelector.kt:70)
Suwayomi-Server  | 	at okhttp3.internal.connection.RealRoutePlanner.planConnect$okhttp(RealRoutePlanner.kt:164)
Suwayomi-Server  | 	at okhttp3.internal.connection.RealRoutePlanner.plan(RealRoutePlanner.kt:75)
Suwayomi-Server  | 	at okhttp3.internal.connection.FastFallbackExchangeFinder.launchTcpConnect(FastFallbackExchangeFinder.kt:119)
Suwayomi-Server  | 	at okhttp3.internal.connection.FastFallbackExchangeFinder.find(FastFallbackExchangeFinder.kt:62)
Suwayomi-Server  | 	at okhttp3.internal.connection.RealCall.initExchange$okhttp(RealCall.kt:298)
Suwayomi-Server  | 	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:32)
Suwayomi-Server  | 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:126)
Suwayomi-Server  | 	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:101)
Suwayomi-Server  | 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:126)
Suwayomi-Server  | 	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:85)
Suwayomi-Server  | 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:126)
Suwayomi-Server  | 	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:74)
Suwayomi-Server  | 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:126)
Suwayomi-Server  | 	at eu.kanade.tachiyomi.network.interceptor.RateLimitInterceptor.intercept(RateLimitInterceptor.kt:123)
Suwayomi-Server  | 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:126)
Suwayomi-Server  | 	at eu.kanade.tachiyomi.network.interceptor.CloudflareInterceptor.intercept(CloudflareInterceptor.kt:44)
Suwayomi-Server  | 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:126)
Suwayomi-Server  | 	at eu.kanade.tachiyomi.network.interceptor.UserAgentInterceptor.intercept(UserAgentInterceptor.kt:21)
Suwayomi-Server  | 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:126)
Suwayomi-Server  | 	at eu.kanade.tachiyomi.network.interceptor.UncaughtExceptionInterceptor.intercept(UncaughtExceptionInterceptor.kt:18)
Suwayomi-Server  | 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:126)
Suwayomi-Server  | 	at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:226)
Suwayomi-Server  | 	at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:574)
Suwayomi-Server  | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
Suwayomi-Server  | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
Suwayomi-Server  | 	at java.base/java.lang.Thread.run(Unknown Source)
```
Since this says atsu.moehttps, I'm pretty sure this is a malformed url.

### Mihon/Mihon fork version

Suwayomi 2.1.2087

### Android version

Not sure if this is relevant for me

### Other details

As far as I can tell this is not related to issue #13367 as it occurs on entries which do not have multiple scanlators. All chapters that fail in suwayomi load normally in atsumaru itself, so it is not a source issue. There does not seem to be any difference in how the images are being stored/encoded between chapters that do/don't fail, and with my limited web programming knowledge I can't find any sort of pattern for which chapters are/aren't working. I asked on the suwayomi discord and according to them it is definitely an extension issue, not a suwayomi issue.

### Acknowledgements

- [x] I have searched [existing issues](https://github.com/keiyoushi/extensions-source/issues) both open & closed, and confirm that this is a new unreported issue.
- [x] I have written a short but informative title.
- [x] I have updated the app to version **[0.19.3](https://github.com/mihonapp/mihon/releases/latest)**.
- [x] I have updated the extension to the latest version (as listed on the [extensions listing](https://keiyoushi.github.io/extensions/) page).
- [x] I have tried the [troubleshooting guide](https://mihon.app/docs/guides/troubleshooting/).
- [x] If this is an issue with the app itself, I should be opening an issue in the app repository.
- [x] I will correctly fill out all of the requested information in this form.

### <!-- footer -->

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/keiyoushi/extensions-source/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/src/en/atsumaru/build.gradle
+++ b/src/en/atsumaru/build.gradle
@@ -1,7 +1,7 @@
 ext {
     extName = 'Atsumaru'
     extClass = '.Atsumaru'
-    extVersionCode = 9
+    extVersionCode = 10
     isNsfw = true
 }
 

--- a/src/en/atsumaru/src/eu/kanade/tachiyomi/extension/en/atsumaru/Atsumaru.kt
+++ b/src/en/atsumaru/src/eu/kanade/tachiyomi/extension/en/atsumaru/Atsumaru.kt
@@ -172,37 +172,24 @@ class Atsumaru : HttpSource() {
 
     // ============================== Chapters ==============================
 
-    private fun fetchChaptersRequest(mangaId: String, page: Int): Request = GET("$baseUrl/api/manga/chapters?id=$mangaId&filter=all&sort=desc&page=$page", apiHeaders)
-
-    override fun chapterListRequest(manga: SManga): Request = fetchChaptersRequest(manga.url, 0)
+    override fun chapterListRequest(manga: SManga): Request = GET("$baseUrl/api/manga/allChapters?mangaId=${manga.url}", apiHeaders)
 
     override fun chapterListParse(response: Response): List<SChapter> {
-        val mangaId = response.request.url.queryParameter("id")!!
+        val mangaId = response.request.url.queryParameter("mangaId")!!
 
-        val mangaPage = try {
+        val scanlatorMap = try {
             val detailsRequest = mangaDetailsRequest(SManga.create().apply { url = mangaId })
             client.newCall(detailsRequest).execute().use {
-                it.parseAs<MangaObjectDto>().mangaPage
-            }
+                it.parseAs<MangaObjectDto>().mangaPage.scanlators?.associate { it.id to it.name }
+            }.orEmpty()
         } catch (_: Exception) {
-            null
+            emptyMap()
         }
 
-        val scanlatorMap = mangaPage?.scanlators?.associate { it.id to it.name }.orEmpty()
-        val chapterToScanlator = mangaPage?.chapters?.associate { it.id to it.scanlationMangaId }.orEmpty()
+        val data = response.parseAs<AllChaptersDto>()
 
-        val chapters = mutableListOf<ChapterDto>()
-        var result = response.parseAs<ChapterListDto>()
-        chapters.addAll(result.chapters)
-
-        while (result.hasNextPage()) {
-            result = client.newCall(fetchChaptersRequest(mangaId, result.page + 1)).execute().parseAs()
-            chapters.addAll(result.chapters)
-        }
-
-        return chapters.map {
-            val scanlatorId = it.scanlationMangaId ?: chapterToScanlator[it.id]
-            it.toSChapter(mangaId, scanlatorId?.let { id -> scanlatorMap[id] })
+        return data.chapters.map {
+            it.toSChapter(mangaId, it.scanlationMangaId?.let { id -> scanlatorMap[id] })
         }
     }
 

--- a/src/en/atsumaru/src/eu/kanade/tachiyomi/extension/en/atsumaru/Atsumaru.kt
+++ b/src/en/atsumaru/src/eu/kanade/tachiyomi/extension/en/atsumaru/Atsumaru.kt
@@ -208,7 +208,12 @@ class Atsumaru : HttpSource() {
     }
 
     override fun pageListParse(response: Response): List<Page> = response.parseAs<PageObjectDto>().readChapter.pages.mapIndexed { index, page ->
-        Page(index, imageUrl = baseUrl + page.image)
+        val imageUrl = when {
+            page.image.startsWith("http") -> page.image
+            page.image.startsWith("//") -> "https:$page.image"
+            else -> "$baseUrl/static/$page.image"
+        }
+        Page(index, imageUrl = imageUrl)
     }
 
     override fun imageRequest(page: Page): Request {

--- a/src/en/atsumaru/src/eu/kanade/tachiyomi/extension/en/atsumaru/Atsumaru.kt
+++ b/src/en/atsumaru/src/eu/kanade/tachiyomi/extension/en/atsumaru/Atsumaru.kt
@@ -178,17 +178,32 @@ class Atsumaru : HttpSource() {
 
     override fun chapterListParse(response: Response): List<SChapter> {
         val mangaId = response.request.url.queryParameter("id")!!
-        val chapterList = mutableListOf<ChapterDto>()
 
+        val mangaPage = try {
+            val detailsRequest = mangaDetailsRequest(SManga.create().apply { url = mangaId })
+            client.newCall(detailsRequest).execute().use {
+                it.parseAs<MangaObjectDto>().mangaPage
+            }
+        } catch (_: Exception) {
+            null
+        }
+
+        val scanlatorMap = mangaPage?.scanlators?.associate { it.id to it.name }.orEmpty()
+        val chapterToScanlator = mangaPage?.chapters?.associate { it.id to it.scanlationMangaId }.orEmpty()
+
+        val chapters = mutableListOf<ChapterDto>()
         var result = response.parseAs<ChapterListDto>()
-        chapterList.addAll(result.chapters)
+        chapters.addAll(result.chapters)
 
         while (result.hasNextPage()) {
             result = client.newCall(fetchChaptersRequest(mangaId, result.page + 1)).execute().parseAs()
-            chapterList.addAll(result.chapters)
+            chapters.addAll(result.chapters)
         }
 
-        return chapterList.map { it.toSChapter(mangaId) }
+        return chapters.map {
+            val scanlatorId = it.scanlationMangaId ?: chapterToScanlator[it.id]
+            it.toSChapter(mangaId, scanlatorId?.let { id -> scanlatorMap[id] })
+        }
     }
 
     override fun getChapterUrl(chapter: SChapter): String {

--- a/src/en/atsumaru/src/eu/kanade/tachiyomi/extension/en/atsumaru/Dto.kt
+++ b/src/en/atsumaru/src/eu/kanade/tachiyomi/extension/en/atsumaru/Dto.kt
@@ -120,13 +120,9 @@ class MangaDto(
 }
 
 @Serializable
-class ChapterListDto(
+class AllChaptersDto(
     val chapters: List<ChapterDto>,
-    val pages: Int,
-    val page: Int,
-) {
-    fun hasNextPage(): Boolean = page + 1 < pages
-}
+)
 
 @Serializable
 class ChapterDto(

--- a/src/en/atsumaru/src/eu/kanade/tachiyomi/extension/en/atsumaru/Dto.kt
+++ b/src/en/atsumaru/src/eu/kanade/tachiyomi/extension/en/atsumaru/Dto.kt
@@ -75,7 +75,13 @@ class MangaDto(
     fun toSManga(baseUrl: String): SManga = SManga.create().apply {
         url = id
         title = this@MangaDto.title
-        thumbnail_url = getImagePath().let { it -> "$baseUrl/static/$it" }
+        thumbnail_url = getImagePath()?.let {
+            when {
+                it.startsWith("http") -> it
+                it.startsWith("//") -> "https:$it"
+                else -> "$baseUrl/static/$it"
+            }
+        }
         description = synopsis
         genre = buildList {
             type?.let { add(it) }

--- a/src/en/atsumaru/src/eu/kanade/tachiyomi/extension/en/atsumaru/Dto.kt
+++ b/src/en/atsumaru/src/eu/kanade/tachiyomi/extension/en/atsumaru/Dto.kt
@@ -59,6 +59,7 @@ class MangaDto(
     private val tags: List<TagDto>? = null,
     private val status: String? = null,
     private val type: String? = null,
+    val scanlators: List<ScanlatorDto>? = null,
 
     // Chapters
     val chapters: List<ChapterDto>? = null,
@@ -110,6 +111,12 @@ class MangaDto(
     class AuthorDto(
         val name: String,
     )
+
+    @Serializable
+    class ScanlatorDto(
+        val id: String,
+        val name: String,
+    )
 }
 
 @Serializable
@@ -123,15 +130,17 @@ class ChapterListDto(
 
 @Serializable
 class ChapterDto(
-    private val id: String,
+    val id: String,
     private val number: Float,
     private val title: String,
+    val scanlationMangaId: String? = null,
     @SerialName("createdAt") private val date: JsonElement? = null,
 ) {
-    fun toSChapter(slug: String): SChapter = SChapter.create().apply {
+    fun toSChapter(slug: String, scanlatorName: String? = null): SChapter = SChapter.create().apply {
         url = "$slug/$id"
         chapter_number = number
         name = title
+        scanlator = scanlatorName
         date?.let {
             date_upload = parseDate(it)
         }

